bokehjs/src/lib/core/util/templating.ts

```diff
@@ -141,15 +148,148 @@ export function _get_column_value(name: string, data_source: ColumnarDataSource,
   }
 }
 
-type PlaceholderType = "$" | "@"
+type PlaceholderType = "$" | "@" | "@$"
 
-export function get_value(type: PlaceholderType, name: string, data_source: ColumnarDataSource, i: Index | null, special_vars: Vars) {
+export function get_value(type: PlaceholderType, name: string, data_source: ColumnarDataSource, index: Index | null, vars: Vars) {
   switch (type) {
-    case "$": return _get_special_value(name, special_vars)
-    case "@": return _get_column_value(name, data_source, i)
+    case "$":  return _get_special_value(name, vars)
+    case "@":  return _get_column_value(name, data_source, index)
+    case "@$": return name == "name" && isString(vars.name) ? _get_column_value(vars.name, data_source, index) : null
   }
 }
 
+class HTML {
+  constructor(public html: string) {}
+}
+
+const functions = {
+  safe: (value: unknown, ...args: unknown[]) => {
+    assert(args.length == 0)
+    if (value == null) {
+      return MISSING
+    } else if (isNumber(value) && isNaN(value)) {
+      return "NaN"
+    } else {
+      return new HTML(`${value}`)
+    }
+  },
+  /*
+  fixed: (value: unknown, ...args: unknown[]) => {
+    assert(args.length == 1)
+    const [digits] = args
+    if (isNumber(value) && isInteger(digits)) {
+      return value.toFixed(digits)
+    } else {
+      return value
+    }
+  },
+  round: (value: unknown, ...args: unknown[]) => {
+    assert(args.length == 0)
+    if (isNumber(value)) {
+      return Math.round(value)
+    } else {
+      return value
+    }
+  },
+  upper: (value: unknown, ...args: unknown[]) => {
+    assert(args.length == 0)
+    if (isString(value)) {
+      return value.toUpperCase()
+    } else {
+      return value
+    }
+  },
+  lower: (value: unknown, ...args: unknown[]) => {
+    assert(args.length == 0)
+    if (isString(value)) {
+      return value.toLowerCase()
+    } else {
+      return value
+    }
+  },
+  filter: (value: unknown, ...args: unknown[]) => {
+    assert(args.length == 1)
+    const [expr] = args
+    if (isPlainObject(expr) && "lit" in expr) {
+      if (expr.lit == "finite") {
+        if (!isNumber(value) || !isFinite(value)) {
+          throw new Skip()
+        }
+      }
+    }
+    return value
+  },
+  */
+}
+
+export function replace_placeholders_html(input: string, data_source: ColumnarDataSource,
+    index: Index | null, formatters?: Formatters, special_vars: Vars = {}): Node[] {
+
+  const html = process_placeholders(input, (type, name, format, _, spec) => {
```

Comment:
It doesn't make sense to factor this part out?

Reply:
This part is specific to handle HTML. Other types of placeholder processors as still plaint text oriented.